### PR TITLE
Scaffolding for meta tensor crossref testing

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75994

There's some utility stuff I could have stacked separately but haven't done so, shout if you want me to.

* `torch.overrides.resolve_name`: this takes a public Torch API function and returns a string name corresponding to it. If found this pretty useful for giving good messages because the repr on most of our function objects is pretty useless (NB: this should be fixed.) It's missing a little bit of functionality; direct calls to torch.ops, torch._VF and torch._C._nn don't work yet. This needed a bit of surgery on `_get_overridable_functions` and a nicer refactor would be good, not exactly sure how to do it.
* I added two utility functions that I needed for test scaffolding: `torch._C. _set_storage_via_tensor`, which is like `set_` but it takes in a tensor rather than a storage (I need this because meta storages don't work), and `torch._C. _is_batched`, which lets me test if a tensor is batched (as far as I can tell, there's no way to test this in userland.)
* I added a new testing envvar `PYTORCH_TEST_WITH_COVERAGE_DB` which will write out a sqlite3 database recording which tests called which torch API functions. This is useful if you're working on a particular torch function and want to quickly execute all tests that exercise that function (not just the obvious ones). From experimentation, I observed that it's best to have a fairly normalized database representation to reduce the disk size and inserts to the database go faster; I define a view to recover the 'naive' viewpoint on the database.

How does crossref testing work? When `PYTORCH_TEST_WITH_CROSSREF=1`, we install a torch function mode which will attempt to run the equivalent of any torch API call with all meta arguments.  For now, we just run it and don't check if any of the results are right, but that is the next logical step.

Doing the same operation, but with meta arguments, is quite involved.

* Input tensors may share the same storage; this must be preserved in the meta tensors so that they share the same storage too; the `meta_storage` ensures if we see the same storage multiple times we map it to the same meta storage. However, meta tensors don't actually support Python storage bindings right now, so I instead return a tensor and use `_set_storage_via_tensor` to install it later.
* The obvious things we have to preserve in the meta tensor are dtype and size. However, here are some more non-obvious things we have to preserve: inference mode, leaf-ness, view-ness, strides, storage offset, conj bit, neg bit.
* Many features we do not support as meta tensors: sparse CSR, sparse, MKLDNN, quantized, nested, functionalization, negative bit, conjugate bit, complex tensors. These don't get transformed, and if a function has an argument that is any of these, we skip the cross-reference test.
* I played around with what to do with tensor subclasses, but in the end I decided that it was too difficult to understand what any given subclass was doing, so we skip cross-ref tests if the tensor in question is not exactly a Tensor or Parameter
* There is a giant blocklist of functions which are known not to work with meta tensors. Previously, I automatically determined this by looking for a NotImplementedError, but this made the test suite run very slowly (cuz C++/Python exceptions are slow). So instead I ran the test suite once slowly, collecting all the functions that failed, and then put them into the blacklist. When you add meta function support, you just delete functions from this list.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

**Comment by ezyang** [comment]
As a fun note, dtype/shape testing found a legit bug in aminmax's meta implementation!

File: torch/overrides.py [hunk 1]
```diff
@@ -1580,6 +1577,10 @@ def get_overridable_functions() -> Dict[Any, List[Callable]]:
                 continue
 
             if not callable(func) and hasattr(func, "__get__"):
+                index[func.__get__] = f"{namespace_str}.{func_name}.__get__"
```
Comment (albanD) [inline torch/overrides.py:1580]: Should this be added only if we actually add this to overridable_funcs below?

File: torch/testing/_internal/common_utils.py [hunk 2]
```diff
@@ -24,6 +24,7 @@
 import contextlib
 import shutil
 import threading
+import sqlite3
```
Comment (albanD) [inline torch/testing/_internal/common_utils.py:27]: Is that a dependency of our test suite? This most likely need a lazy import.

File: torch/testing/_internal/common_utils.py [hunk 3]
```diff
@@ -807,10 +861,817 @@ def wrapper(*args, **kwargs):
...
+        self.test_case = test_case
+
     def __torch_function__(self, func, types, args=(), kwargs=None):
         kwargs = kwargs or {}
+
+        hit = 0
+        miss = 0
+
+        # Doesn't actually return a storage
+        @functools.lru_cache(None)
```
Comment (albanD) [inline torch/testing/_internal/common_utils.py:1538]: We call this a lot with the same Storage?

File: torch/testing/_internal/common_utils.py [hunk 4]
```diff
@@ -807,10 +861,817 @@ def wrapper(*args, **kwargs):
...
+                # inference mode can trigger this
+                return False
+
+        @functools.lru_cache(None)
+        def meta_tensor(t):
+            with torch.inference_mode(t.is_inference()):
+                s = meta_storage(t.storage())
+                is_leaf = safe_is_leaf(t)
+                if is_leaf or not t._is_view():
+                    r = torch.empty(
```
Comment (albanD) [inline torch/testing/_internal/common_utils.py:1555]: We really need to expose the `new_empty_strided` function!

File: torch/testing/_internal/common_utils.py [hunk 5]
```diff
@@ -807,10 +861,817 @@ def wrapper(*args, **kwargs):
...
+                        if t._is_view() and not safe_is_leaf(t._base):
+                            base = base.clone()
+                        r = base.as_strided(t.size(), t.stride(), t.storage_offset())
+                torch._C._set_conj(r, t.is_conj())
+                torch._C._set_neg(r, t.is_neg())
+            return r
+
+        def to_meta(t):
+            nonlocal hit, miss
+            # TODO: zero tensors?  We appear to have eliminated them by
```
Comment (albanD) [inline torch/testing/_internal/common_utils.py:1583]: Zero Tensors only appear in the forward AD for now (unless you do fun stuff). And all of that is in c++ so that sounds ok.
We could add a test with Custom Function that would trigger that though I think.

File: torch/testing/_internal/common_utils.py [hunk 6]
```diff
@@ -807,10 +861,817 @@ def wrapper(*args, **kwargs):
...
+                    )
+                    torch._C._set_storage_via_tensor(
+                        base, s, 0, s.size(), (1,)
+                    )
+                    base.requires_grad = t.requires_grad
+                    with torch.enable_grad():
+                        if t._is_view() and not safe_is_leaf(t._base):
+                            base = base.clone()
+                        r = base.as_strided(t.size(), t.stride(), t.storage_offset())
+                torch._C._set_conj(r, t.is_conj())
```
Comment (albanD) [inline torch/testing/_internal/common_utils.py:1577]: Should we remove this to avoid confusion since we skip these below?

File: torch/testing/_internal/common_utils.py [hunk 7]
```diff
@@ -807,10 +861,817 @@ def wrapper(*args, **kwargs):
...
+                        suppress = True
+                if not suppress:
+                    raise RuntimeError(f"""\
+failed to run: {torch.overrides.resolve_name(func) or func}(
+    *{meta_args},
+    **{meta_kwargs}
+  )""") from e
+            else:
+                pass
+                # self.test_case.assertEqual(meta_r.dtype, r.dtype)
```
Comment (albanD) [inline torch/testing/_internal/common_utils.py:1673]: Not passing?

**Review by albanD — COMMENTED** [review]
Looks great!
I would have expected more test to fail actually!

File: torch/testing/_internal/common_utils.py [hunk 8]
```diff
@@ -807,10 +861,818 @@ def wrapper(*args, **kwargs):
...
+        )
+
+        if do_meta:
+            try:
+                meta_args = tree_map(to_meta, args)
+                meta_kwargs = tree_map(to_meta, kwargs)
+            except Exception as e:
+                raise RuntimeError(
+                    f"failed to convert args to meta; "
+                    f"originally (*{args}, **{kwargs})") from e
```
Comment (zou3519) [inline torch/testing/_internal/common_utils.py:1642]: High-level question: is crossref meant to test only meta, or will it also test other things (like torch-> prim decompositions, aten->??? decompositions) in the future?

If it is the latter: right now there is a single skipIfCrossRef decorator, we would want different decorators for different type of crossref tests

File: torch/csrc/Module.cpp [hunk 9]
```diff
@@ -1067,6 +1068,13 @@ Call this whenever a new thread is created in order to propagate values from
   py_module.def("_set_neg", [](const at::Tensor & x, bool neg) {
     x._set_neg(neg);
   });
+  py_module.def("_set_storage_via_tensor", [](const at::Tensor & x, const at::Tensor & y, int64_t storage_offset, at::IntArrayRef size, at::IntArrayRef stride) {
+    TORCH_CHECK(y.is_contiguous());
+    x.set_(y.storage(), storage_offset, size, stride);
+  });
+  py_module.def("_is_batched", [](const at::Tensor & x) {
+    return at::isBatchedTensor(x);
+  });
```
Comment (zou3519) [inline torch/csrc/Module.cpp:1077]: Oh no, what are you using this for? (still making my way through)

Comment (ezyang) [reply torch/csrc/Module.cpp:1077]: It's to detect if an input is a batched tensor, and then skip the crossref in that case. I actually think that we should be able to test this case, but I need to detect that something is a batched tensor and then reconstruct the same thing, but meta (but with appropriate storage/strides so I can't just call `to("meta")`). It feels like I don't have enough API surface but I also didn't look that hard.

File: torch/testing/_internal/common_utils.py (hunk 8 above)
Comment (ezyang) [reply torch/testing/_internal/common_utils.py:1642]: Yes, this is probably right. Some of the tests I'm skipping probably should be permanently skipped for all cross refs because they're very painful (cpp extensions sticks out). Actually, some of these skips may be possible to remove again since I tightened cross ref conditions

File: torch/testing/_internal/common_utils.py (hunk 7 above)
Comment (ezyang) [reply torch/testing/_internal/common_utils.py:1674]: Gonna turn these on later with fixes, but yes, autocast + meta is broken and fails this assert, because autocast rules are per device.

File: torch/testing/_internal/common_utils.py [hunk 10]
```diff
@@ -807,10 +861,817 @@ def wrapper(*args, **kwargs):
...
+                ]):
+                    # TODO: sparse should support meta
+                    # NB technically to('meta') does work but our logging
+                    # instrumentation will see the meta conversions and the
+                    # tests all break so we just exclude this.  In any case
+                    # the to conversion isn't really right anyhow.
+                    miss += 1
+                    return t
+                elif any([
+                    t.device.type in ("lazy", "meta"), t.is_complex(),
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/testing/_internal/common_utils.py:1604]: huh, odd, I swear I had to add the complex dtype tests later because there were problems :/ Will check

File: torch/csrc/Module.cpp [hunk 11]
```diff
@@ -1067,6 +1068,13 @@ Call this whenever a new thread is created in order to propagate values from
   py_module.def("_set_neg", [](const at::Tensor & x, bool neg) {
     x._set_neg(neg);
   });
+  py_module.def("_set_storage_via_tensor", [](const at::Tensor & x, const at::Tensor & y, int64_t storage_offset, at::IntArrayRef size, at::IntArrayRef stride) {
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/csrc/Module.cpp:1071]: agreed

File: test/test_autograd.py [hunk 12]
```diff
@@ -6191,6 +6192,8 @@ def test_setting_default_saved_variable_hooks_twice_should_use_inner(self):
         self.assertEqual(2 * 5 * 5 * a, a.grad)
         self.assertEqual(2 * 3 * 3 * b, b.grad)
 
+    # this is some crazy pants, IDK how to do this with mode
```
[the remark this answers is not in the document]
Comment (ezyang) [reply test/test_autograd.py:6195]: I got some really weird error message with the mode test involving the grad fns. At the time, I thought there was some sort of serialization/deserialization, although on reflection here that may not be what happened. Will bring up the error message here.

File: torch/csrc/Module.cpp [hunk 13]
```diff
@@ -1067,6 +1068,13 @@ Call this whenever a new thread is created in order to propagate values from
...
+    TORCH_CHECK(y.is_contiguous());
+    x.set_(y.storage(), storage_offset, size, stride);
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/csrc/Module.cpp:1073]: yup

File: torch/csrc/utils/python_arg_parser.cpp [hunk 14]
```diff
@@ -286,7 +286,7 @@ auto handle_torch_function_no_python_arg_parser(
       // See https://github.com/pytorch/pytorch/issues/63767
       if (PyObject_FastGetAttrString(torch_function.ptr(), "__self__").is(arg) &&
           torch_function.ptr() != torch::disabled_torch_function_impl()) {
-        TORCH_WARN("Defining your `__torch_function__` as a plain method is deprecated and ",
+        TORCH_WARN_ONCE("Defining your `__torch_function__` as a plain method is deprecated and ",
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/csrc/utils/python_arg_parser.cpp:289]: Some of the tests were spamming this a lot so I added this to shut it up

File: torch/overrides.py [hunk 15]
```diff
@@ -1531,36 +1532,32 @@ def handle_torch_function(
...
             # ignore private functions or functions that are deleted in torch.__init__
             if namespace is not torch.Tensor:
-                if func_name.startswith('_'):
+                if func_name.startswith('__'):
                     continue
+                elif func_name.startswith('_'):
+                    ignore = True
                 elif func_name.endswith('_'):
-                    continue
+                    ignore = True
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/overrides.py:1558]: That's what the old code did. For resolving I don't ignore these.

File: torch/overrides.py (hunk 1 above)
Comment (ezyang) [reply torch/overrides.py:1580]: No, the name resolution and list of overridable funcs is intentionally different (although maybe it shouldn't be). Overridable funcs is not actually a complete list of all the functions that can show up in `__torch_function__`, and for name resolution I want as complete a list as possible

File: torch/testing/_internal/common_utils.py [hunk 16]
```diff
@@ -807,10 +861,817 @@ def wrapper(*args, **kwargs):
...
+        test_file = os.path.relpath(full_test_file, prefix)
+    else:
+        test_file = '<stdin>'
+
+    def __init__(self, conn, test_case):
+        self.conn = conn
+        self.test_case = test_case
+        self.seen = set()
+
+    def __torch_function__(self, func, types, args=(), kwargs=None):
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/testing/_internal/common_utils.py:1507]: torch function modes are not class methods, this is intentional :)

File: torch/testing/_internal/common_utils.py (hunk 3 above)
Comment (ezyang) [reply torch/testing/_internal/common_utils.py:1539]: not really, but it's important to give the same tensor back if you hit the same storage

File: torch/testing/_internal/common_utils.py [hunk 17]
```diff
@@ -807,10 +861,817 @@ def wrapper(*args, **kwargs):
...
+        @functools.lru_cache(None)
+        def meta_storage(s):
+            return torch.empty(s.size(), dtype=s.dtype, device='meta')
+
+        def safe_is_leaf(t):
+            try:
+                return t.is_leaf
+            except RuntimeError:
+                # inference mode can trigger this
+                return False
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/testing/_internal/common_utils.py:1548]: here's how I triggered it

```
import torch

s = torch.ones(1, 2, 3, requires_grad=True)
a = s.clone()

with torch.inference_mode():
    out = a.view(-1)
    tmp = out.view(-1)
    tmp.add_(2)
    tmp.is_leaf
```

File: torch/testing/_internal/common_utils.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (albanD) [reply torch/testing/_internal/common_utils.py:1507]: This is most likely why you get so many warnings? Maybe we want to review that policy again?

File: torch/testing/_internal/common_utils.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (albanD) [reply torch/testing/_internal/common_utils.py:1548]: That is expected. The inplace op there should raise an error. And it is only because of the lazyness of view that we only realize later.

[the remark this answers is not in the document]
Comment (ezyang) [reply torch/testing/_internal/common_utils.py:1548]: And therefore I need safe_is_leaf. TBH I'm not very happy with the property access causing an exception.

[the remark this answers is not in the document]
Comment (albanD) [reply torch/testing/_internal/common_utils.py:1548]: Accessing `.grad_fn` or calling `.backward()` (anything updating the grad_fn really) will raise an error.

File: torch/testing/_internal/common_utils.py [hunk 18]
```diff
@@ -601,6 +607,53 @@ def run_tests(argv=UNITTEST_ARGS):
...
+        global COVERAGE_DB_FILE_ID
+        con = sqlite3.connect(TEST_WITH_COVERAGE_DB, isolation_level=None)
+
+        con.executescript("""
+PRAGMA journal_mode = OFF;
+PRAGMA synchronous = 0;
+PRAGMA locking_mode = EXCLUSIVE;
+PRAGMA temp_store = MEMORY;
+
+CREATE TABLE IF NOT EXISTS files ( file_id INTEGER NOT NULL PRIMARY KEY, file TEXT NOT NULL UNIQUE );
```
Comment (eellison) [inline torch/testing/_internal/common_utils.py:621]: never thought id see SQL in pytorch 😮

**Review by eellison — COMMENTED** [review]
What is the plan for this PR? Context is OpInfos dont have any mixed-device inputs so they're not very useful for testing FakeTensors, you would need something like this instead

**Comment by ezyang** [comment]
I'm currently not sure how to prioritize it. You get very good coverage with it, but it's also very painful to do development on: it takes a few hours to run through the entirety of the test suite and when you issue fixes it is somewhat difficult to understand which tests you should rerun to revalidate. Without fake tensors is mind, I was planning on getting the OpInfo test suite clean first, and then moving on to crossref testing.

Here is my suggestion, @eellison. You should create a variant mode that looks for tests that exercise mixed device inputs and log them all. Then (somehow) selectively apply cross-ref testing on this set (the somehow because I don't know how to automatically apply something to a generated list of tests, you'll need to figure something out) and land that.
